tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition.py

```diff
@@ -0,0 +1,517 @@
+        visited.add(input_node_name)
+        input_node = node_name_to_node_def[input_node_name]
+
+        # Stop traversing when reaching a remote op.
+        if _is_remote_op(input_node):
+          remote_op_immediate_dep.append(input_node_name)
+        else:
+          queue.append(input_node_name)
+
+  return remote_op_immediate_dep
+
+
+def _is_placeholder_op(node: node_def_pb2.NodeDef) -> bool:
+  return node.op == "Placeholder"
+
+
+def _is_remote_op(node: node_def_pb2.NodeDef) -> bool:
+  return node.op == "PyFunc"
+
+
+def _get_execution_specs(
+    graph_def: graph_pb2.GraphDef,
+    graph_output_names: List[Text],
+    graph: tf.Graph,
+    node_name_to_node_def: Mapping[Text, node_def_pb2.NodeDef],
+    remote_op_to_immediate_dep: Mapping[Text, List[Text]]
+) -> List[ExecutionSpec]:
+  """Generates the ExecutionSpecs for a graph.
+
+  A "layer" contains one or more nodes inside a graph. There are two types of
+  layers: subgraph layer and remote op layer. A subgraph layer doesn't
+  contain remote ops, whereas a remote op layer only contains remote ops.
+  Remote ops inside a remote op layer don't depend on each other's output,
+  so the order of execution between those remote ops doesn't matter.
+
+  We first identify the remote op layers of a graph. Then, based on the
+  remote op layers, we can derive the subgraph layers. For example, after
+  identifying the first remote op layer, we can equate the inputs of the
+  remote op layer to the outputs of the previous subgraph layer. We can
+  then traverse and construct the previous subgraph layer.
+
+  Each subgraph layer can be captured into one ExecutionSpec, but each
+  remote op layer need to be stored into N ExecutionSpecs, where N equals
+  to the number of remote ops inside a remote op layer. This happens
+  because each remote op essentially represents a graph.
+
+  Args:
+    graph_def: A `GraphDef` proto.
+    graph_output_names: A list of graph output node names.
+    graph: A tf.Graph representing the same graph as graph_def.
+    node_name_to_node_def: A mapping from node names to `NodeDef` protos.
+    remote_op_to_immediate_dep: A mapping from remote op name to a list of
+                                remote op immediate dependencies' names.
+
+  Returns:
+    A list of ExecutionSpecs, where the order of the list represents the
+    order of execution.
+  """
+  execution_specs = []  # type: List[ExecutionSpec]
+  previously_visited = set()  # type: Set[Text]
+
+  for remote_op_layer in _RemoteOpLayers(remote_op_to_immediate_dep):
+    # Get one subgraph layer
+    output_node_names = _get_previous_subgraph_layer_output_node_names(
+        remote_op_layer, node_name_to_node_def)
+
+    if output_node_names:
+      spec = _get_execution_spec_for_subgraph_layer(
+          graph_def, graph, node_name_to_node_def,
+          previously_visited, output_node_names)
+      execution_specs.append(spec)
+      previously_visited |= _get_non_input_names(spec.subgraph)
```

Comment:
Does it actually matter whether we get non-input nodes only or all nodes from `spec.subgraph`?

Reply:
No. The implementation checks whether a node is a remote op, a placeholder op, or a node visited in previous layers. Therefore, non-input nodes suffices the purpose.